Basic Auth for pinot-controller

apucher opened the pull request:
## tl;dr for review
`AuthQuickstart.java` (sample setup with basic auth enabled)
`BasicAuthUtils.java` (shared auth logic)
`BasicAuthAccessControlFactory.java` x 2 (broker, controller auth impl)

`authToken` is used (usually) to authenticate to a controller API via an `Authorization` http header
`authToken = null` (no auth)
`authToken = Basic abcdef...` (basic auth)
`authToken = Bearer 012345...` (token / oauth2)

duplicate config properties for auth tokens due to config subsets being passed around. alternatives considered: single global key with injection into different config subsets (possible), single global key injected via singleton provider (anti-pattern)

auth token for tasks are injected by the minion runner directly, if not provided in the task configs (common case)

## Description
Adds support for token-based authentication in pinot-controller, pinot-broker, pinot-server, minion, and any associated tooling. Furthermore, adds endpoints to manage different auth workflows via UI and provides a sample implementation using basic auth.

**New access control factories**
`org.apache.pinot.controller.api.access.BasicAuthAccessControlFactory`
`org.apache.pinot.broker.broker.BasicAuthAccessControlFactory`

Changes are fully backwards-compatible with authentication disabled by default. Enabling basic auth requires future work for UI support. Builds on #6507, relates to #6435.

## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [x] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)

## Release Notes
Adds support for token-based authentication in pinot-controller, pinot-broker, pinot-server, minion, and any associated tooling. Furthermore, adds endpoints to manage different auth workflows via UI and provides a sample implementation using basic auth.

**New endpoints:**

pinot-controller
`GET /auth/verify` (optional query params tableName, accessType, endpointUrl)
returns true if the provided credentials permit access to the given resource, or false otherwise.

`GET /auth/info`
returns an instance of AuthWorkflowInfo, which at minimum provides a workflow name. Currently supports NONE and BASIC and may be extended later. Support for each type from the UI is optional.

**New configuration properties:**

pinot-controller
`controller.segment.fetcher.auth.token`

pinot-server
`pinot.server.segment.fetcher.auth.token`
`pinot.server.segment.upload.auth.token`
`pinot.server.instance.auth.token`

minion
`segment.fetcher.auth.token`
`task.auth.token`

## Documentation
Quickstart with basic auth
`org.apache.pinot.tools.AuthQuickstart`

curl samples for controller endpoints:
`curl localhost:9000/auth/info` -> `{"workflow":"NONE"}`
`curl localhost:9000/auth/info` -> `{"workflow":"BASIC"}`
`curl localhost:9000/auth/info` -> `{"workflow":"OAUTH2", "issuerUrl":"https://xxxxx"}`
`curl localhost:9000/auth/verify` -> `false`
`curl localhost:9000/auth/verify -H 'Authorization: Basic YWRtaW46dmVyeXNlY3JldA=='` -> `true`

Sample configuration properties for basic auth, plus client auth tokens.

pinot-controller
`controller.segment.fetcher.auth.token`=Basic YWRtaW46dmVyeXNlY3JldA

`controller.admin.access.control.factory.class`=org.apache.pinot.controller.api.access.BasicAuthAccessControlFactory
`controller.admin.access.control.principals`=admin,user
`controller.admin.access.control.principals.admin.password`=verysecret
`controller.admin.access.control.principals.user.password`=secret
`controller.admin.access.control.principals.user.tables`=baseballStats
`controller.admin.access.control.principals.user.permissions`=read

pinot-broker
`pinot.broker.access.control.class`=org.apache.pinot.broker.broker.BasicAuthAccessControlFactory
`pinot.broker.access.control.principals`=admin,user
`pinot.broker.access.control.principals.admin.password`=verysecret
`pinot.broker.access.control.principals.user.password`=secret
`pinot.broker.access.control.principals.user.tables`=baseballStats
`pinot.broker.access.control.principals.user.permissions`=read

pinot-server
`pinot.server.segment.fetcher.auth.token`=Basic YWRtaW46dmVyeXNlY3JldA
`pinot.server.segment.uploader.auth.token`=Basic YWRtaW46dmVyeXNlY3JldA
`pinot.server.instance.auth.token`=Basic YWRtaW46dmVyeXNlY3JldA

minion
`segment.fetcher.auth.token`=Basic YWRtaW46dmVyeXNlY3JldA
`task.auth.token`=Basic YWRtaW46dmVyeXNlY3JldA

apucher commented:
@icefury71

icefury71 commented:
Thanks so much for this PR. As part of this can we also add a new API to enable credential verification ? (eg: from the Cluster Manager UI). I'm thinking of something like this: "http://controller-host:port/verifyauth" where you pass in the token and verify whether its correct. 

The other option is to annotate one of the existing APIs eg: GET /tables with "Authenticate" and reuse that. However, a new API might be cleaner. What do you think ?

apucher commented:
> Thanks so much for this PR. As part of this can we also add a new API to enable credential verification ? (eg: from the Cluster Manager UI). I'm thinking of something like this: "http://controller-host:port/verifyauth" where you pass in the token and verify whether its correct.
> 
> The other option is to annotate one of the existing APIs eg: GET /tables with "Authenticate" and reuse that. However, a new API might be cleaner. What do you think ?

added `/auth/info` and `/auth/verify` endpoints.
**info** provides a workflow name to be supported from the web UI, including `NONE`, `BASIC`, and `OAUTH2`
**verify** allows to verify access to API endpoints in a generic fashion, or simply to sanity-check auth headers without additional query params

apucher commented on pinot-common/src/main/java/org/apache/pinot/common/segment/generation/SegmentGenerationUtils.java:
```diff
@@ -87,7 +94,12 @@ public static Schema getSchema(String schemaURIString) {
     }
   }
 
+  @Deprecated
```
users can keep *configuring* nulls, but any dev work should consider authentication from this point on. otherwise, we end up with a nightmare of a compatibility matrix for new features and auth support

icefury71 replied:
Sounds good.

apucher commented on pinot-controller/src/main/java/org/apache/pinot/controller/api/access/BasicAuthAccessControlFactory.java:
```diff
@@ -0,0 +1,106 @@
+
+  @Override
+  public AccessControl create() {
+    return _accessControl;
```
I have to maintain compatibility with the existing interfaces. pinot's config management isn't great, and this seems to be the only graceful way of getting access to pinot's config properties

apucher commented on pinot-core/src/main/java/org/apache/pinot/core/auth/BasicAuthPrincipal.java:
```diff
@@ -0,0 +1,55 @@
+  private final String _name;
+  private final String _token;
+  private final Set<String> _tables;
+  private final Set<String> _permissions;
```
should be `String` to retain extensibility of AccessControl implementations

sajjad-moradi replied:
I think it should be "AccessType" because AccessControl.hasAccess() takes an AccessType object as its argument. Implementations of AccessControl can't change AccessType.

apucher replied:
(Double-posting here for context) I agree with this from the controller side, however, AccessType is a controller concept only - broker implementations may specify other types of access in an extensible way. One could imagine limiting access to PII data and similar

apucher commented on pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotControllerAuthResource.java:
```diff
@@ -0,0 +1,75 @@
+  @Path("auth/verify")
+  @Produces(MediaType.APPLICATION_JSON)
+  @ApiOperation(value = "Check whether authentication is enabled")
+  @ApiResponses(value = {@ApiResponse(code = 200, message = "Verification result provided"), @ApiResponse(code = 500, message = "Verification error")})
```
`/auth/verify` does not *require* authentication itself, i.e. it can't be `Forbidden`. it does verify a provided token and return 200 with a body (true/false) or a 500 if there's an error processing the token

icefury71 replied:
Makes sense

apucher commented on pinot-controller/src/main/java/org/apache/pinot/controller/api/access/BasicAuthAccessControlFactory.java:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    @Override
+    public boolean hasAccess(String tableName, AccessType accessType, HttpHeaders httpHeaders, String endpointUrl) {
```
err ... `hasAccess()`, etc. are defined in the `AccessControl` interface and `@Override` ensures a method signature match

icefury71 replied:
My bad - I looked at the wrong interface. Sorry for the noise

apucher commented on pinot-controller/src/main/java/org/apache/pinot/controller/api/access/BasicAuthAccessControlFactory.java:
```diff
@@ -0,0 +1,106 @@
+      this._principals = principals.stream().collect(Collectors.toMap(BasicAuthPrincipal::getToken, p -> p));
+    }
+
+    @Override
```
https://stackoverflow.com/questions/94361/when-do-you-use-javas-override-annotation-and-why

apucher commented on pinot-controller/src/main/java/org/apache/pinot/controller/api/access/BasicAuthAccessControlFactory.java:
```diff
@@ -0,0 +1,106 @@
+    _accessControl = new BasicAuthAccessControl(BasicAuthUtils.extractBasicAuthPrincipals(configuration, PREFIX));
+  }
+
+  @Override
```
not required, but preferred

apucher commented on pinot-controller/src/main/java/org/apache/pinot/controller/api/access/BasicAuthAccessControlFactory.java:
```diff
@@ -0,0 +1,106 @@
+   * Access Control using header-based basic http authentication
+   */
+  private static class BasicAuthAccessControl implements AccessControl {
+    private final Map<String, BasicAuthPrincipal> _principals;
```
fixed

mcvsubbu commented on pinot-common/src/main/java/org/apache/pinot/common/utils/fetcher/SegmentFetcherFactory.java:
```diff
@@ -34,29 +34,40 @@
-  private static final SegmentFetcher DEFAULT_HTTP_SEGMENT_FETCHER = new HttpSegmentFetcher();
-  private static final SegmentFetcher DEFAULT_PINOT_FS_SEGMENT_FETCHER = new PinotFSSegmentFetcher();
-
-  static {
```
Need to look at the history of this class carefully. The last time we made some changes like this, we had issues where the controller (I think) had an exception, probably during startup. It was when this whole thing was in com.linkedin space, so we don't have the history readily available.

apucher replied:
The use of this class is pretty messy. The fetchers here were instantiated statically without configuration, but we now may need to inject authTokens. I've therefore refactored the class to maintain its public (static) interface, while setting properties in the starter before use.
With my manual testing, quickstart, and integration tests things seem to work, but I bet there;s a corner case lurking somewhere.

mcvsubbu replied:
I examined the history dating back to PRs #1970 (caused a problem), #1973, #2171, #2200, #2210, and #2238. I am not able to spot anything that may cause problems. If the code structure will allow it,  if you can commit this class independently that would be awesome. If not, we will take this.

mcvsubbu commented on pinot-common/src/main/java/org/apache/pinot/common/utils/fetcher/SegmentFetcherFactory.java:
```diff
@@ -34,29 +34,40 @@
+
+  private void initInternal(PinotConfiguration config)
+      throws Exception {
+    _httpSegmentFetcher.init(config); // directly, without sub-namespace
```
Am I missing this? Are these added to the map anywhere as default fetchers?

apucher replied:
these replace the static fetcher instances from the comment above

mcvsubbu commented on pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotControllerAuthResource.java:
```diff
@@ -0,0 +1,75 @@
+  }
+
+  @GET
+  @Path("auth/info")
```
what is auth workflow?

apucher replied:
added javadoc. auth info help the Pinot UI configure different auth workflows - no auth, basic auth, and oauth2

mcvsubbu reviewed (commented):
high level review, not complete

apucher commented on pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java:
```diff
@@ -180,6 +180,16 @@ public BrokerResponse handleRequest(JsonNode request, @Nullable RequesterIdentit
 
+    // first-stage access control to prevent unauthenticated requests from using up resources
+    // secondary table-level check comes later
+    boolean hasAccess = _accessControlFactory.create().hasAccess(requesterIdentity, null);
```
added

apucher commented on pinot-common/src/main/java/org/apache/pinot/common/utils/CommonConstants.java:
```diff
@@ -235,6 +235,9 @@
     public static final String PREFIX_OF_CONFIG_OF_PINOT_FS_FACTORY = "pinot.server.storage.factory";
     public static final String PREFIX_OF_CONFIG_OF_PINOT_CRYPTER = "pinot.server.crypter";
+
+    public static final String CONFIG_OF_AUTH_TOKEN = "auth.token";
```
added

apucher commented on pinot-common/src/main/java/org/apache/pinot/common/utils/CommonConstants.java:
```diff
@@ -304,6 +307,7 @@
       public static final String CONFIG_OF_CONTROLLER_HTTPS_ENABLED = "enabled";
       public static final String CONFIG_OF_CONTROLLER_HTTPS_PORT = "controller.port";
       public static final String CONFIG_OF_SEGMENT_UPLOAD_REQUEST_TIMEOUT_MS = "upload.request.timeout.ms";
+      public static final String CONFIG_OF_SEGMENT_UPLOAD_AUTH_TOKEN = "upload.auth.token";
```
added

apucher commented on pinot-common/src/main/java/org/apache/pinot/common/utils/CommonConstants.java:
```diff
@@ -353,6 +357,8 @@
     public static final String PREFIX_OF_CONFIG_OF_SEGMENT_UPLOADER = "segment.uploader";
     public static final String PREFIX_OF_CONFIG_OF_PINOT_CRYPTER = "crypter";
+
+    public static final String CONFIG_OF_TASK_AUTH_TOKEN = "task.auth.token";
```
added

apucher commented on pinot-common/src/main/java/org/apache/pinot/common/utils/FileUploadDownloadClient.java:
```diff
@@ -168,16 +172,6 @@ public static URI getRetrieveSchemaURI(String protocol, String host, int port, S
     return getURI(protocol, host, port, SCHEMA_PATH + "/" + schemaName);
   }
 
-  public static URI getUploadSchemaHttpURI(String host, int port)
```
I've returned the class to its original public interface and flagged any methods as deprecated instead. This should ease the transition and retain full backwards-compatibility

mcvsubbu commented on pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java:
```diff
@@ -180,6 +180,16 @@ public BrokerResponse handleRequest(JsonNode request, @Nullable RequesterIdentit
+    // secondary table-level check comes later
+    boolean hasAccess = _accessControlFactory.create().hasAccess(requesterIdentity);
+    if (!hasAccess) {
+      _brokerMetrics.addMeteredTableValue(null, BrokerMeter.REQUEST_DROPPED_DUE_TO_ACCESS_ERROR, 1);
```
You may want to consider a different metric here than in line 261

apucher replied:
imo this is the same concern. first-stage access control simply limits resource consumption (i.e. denial of service, leakage of non-table data) for unauthenticated requests. second-stage performs in-depth authorization per table. any rejections are are ACL-related.

mcvsubbu replied:
Ah, I know what I wanted to say:
```suggestion
      _brokerMetrics.addMeteredGlobalValue(BrokerMeter.REQUEST_DROPPED_DUE_TO_ACCESS_ERROR, 1);
```

apucher replied:
done

mcvsubbu commented on pinot-common/src/main/java/org/apache/pinot/common/utils/CommonConstants.java:
```diff
@@ -305,6 +312,12 @@
       public static final String CONFIG_OF_SEGMENT_UPLOAD_REQUEST_TIMEOUT_MS = "upload.request.timeout.ms";
 
+      /**
+       * Service token for accessing protected controller APIs.
```
Hmmm what is the difference between this config and the config in line 243 (and then in 371)? The comments all say the same thing :)

apucher replied:
all these do the same thing: authenticate with controller APIs. The only need for their replication is due to pinot's handling of configuration subsets. segment fetcher, uploader, etc. all manage their token separately.

mcvsubbu commented on pinot-common/src/main/java/org/apache/pinot/common/utils/FileUploadDownloadClient.java:
```diff
@@ -139,25 +140,41 @@ private static URI getURI(String protocol, String host, int port, String path)
   }
 
+  /**
+   * Deprecated due to lack of protocol/scheme support. May break for deployments with TLS/SSL enabled
```
Can you provide a pointer to the method to be used instead? Do we need to introduce a method that takes authToken as an additional argument ? Perhaps it should specify "null" (the string) in the calling sequence?

apucher replied:
added

mcvsubbu commented on pinot-common/src/main/java/org/apache/pinot/common/utils/FileUploadDownloadClient.java:
```diff
@@ -139,25 +140,41 @@ private static URI getURI(String protocol, String host, int port, String path)
   }
 
+  /**
+   * Deprecated due to lack of protocol/scheme support. May break for deployments with TLS/SSL enabled
```
same. need an alternative so the user can migrate and stop using the deprecated methods. We can remove the deprecated ones in a subsequent  release.

apucher replied:
added

mcvsubbu commented on pinot-common/src/main/java/org/apache/pinot/common/utils/FileUploadDownloadClient.java:
```diff
@@ -139,25 +140,41 @@ private static URI getURI(String protocol, String host, int port, String path)
+  /**
+   * Deprecated due to lack of protocol/scheme support. May break for deployments with TLS/SSL enabled
+   */
+  @Deprecated
```
same

apucher replied:
added

mcvsubbu commented on pinot-common/src/main/java/org/apache/pinot/common/utils/FileUploadDownloadClient.java:
```diff
@@ -139,25 +140,41 @@ private static URI getURI(String protocol, String host, int port, String path)
   }
 
+  /**
+   * Deprecated due to lack of protocol/scheme support. May break for deployments with TLS/SSL enabled
```
same

apucher replied:
added

mcvsubbu commented on pinot-common/src/main/java/org/apache/pinot/common/utils/FileUploadDownloadClient.java:
```diff
@@ -139,25 +140,41 @@ private static URI getURI(String protocol, String host, int port, String path)
   }
 
+  /**
+   * Deprecated due to lack of protocol/scheme support. May break for deployments with TLS/SSL enabled
```
same

apucher replied:
added

mcvsubbu commented on pinot-common/src/main/java/org/apache/pinot/common/utils/FileUploadDownloadClient.java:
```diff
@@ -168,11 +185,19 @@ public static URI getRetrieveSchemaURI(String protocol, String host, int port, S
   }
 
+  /**
+   * Deprecated due to lack of protocol/scheme support. May break for deployments with TLS/SSL enabled
```
same

apucher replied:
added

mcvsubbu commented on pinot-common/src/main/java/org/apache/pinot/common/utils/fetcher/BaseSegmentFetcher.java:
```diff
@@ -36,6 +35,7 @@
   public static final String RETRY_COUNT_CONFIG_KEY = "retry.count";
   public static final String RETRY_WAIT_MS_CONFIG_KEY = "retry.wait.ms";
   public static final String RETRY_DELAY_SCALE_FACTOR_CONFIG_KEY = "retry.delay.scale.factor";
+  public static final String AUTH_TOKEN = "auth.token";
```
+1 As long as it has the same semantics as the auth token defined there, we should be able to re-use that.

apucher replied:
done

mcvsubbu reviewed (commented):
Do you need to change LLCSegmentCompletionHandlers to pay attention to the auth token?

apucher commented:
> Do you need to change LLCSegmentCompletionHandlers to pay attention to the auth token?

Probably. I'd appreciate your input on where this actually talks to controller APIs. I just noticed that this whole set of endpoints was excluded from my checklist (via swagger)

mcvsubbu commented:
> > Do you need to change LLCSegmentCompletionHandlers to pay attention to the auth token?
> 
> Probably. I'd appreciate your input on where this actually talks to controller APIs. I just noticed that this whole set of endpoints was excluded from my checklist (via swagger)

These endpoints should not appear in Swagger because they are used by pinot-server to commit realtime segments using the segment commit protocol. They are not meant for administrative or user actions. These are completely pinot internal operations.
The client side of this call is in ServerSegmentProtocolCompletionHandler.

If you can make these changes in another PR, that will help reviewing a bit. If I miss something and it causes realtime segments not to complete segments, that will be really bad.

apucher commented:
> > > Do you need to change LLCSegmentCompletionHandlers to pay attention to the auth token?
> > 
> > 
> > Probably. I'd appreciate your input on where this actually talks to controller APIs. I just noticed that this whole set of endpoints was excluded from my checklist (via swagger)
> 
> These endpoints should not appear in Swagger because they are used by pinot-server to commit realtime segments using the segment commit protocol. They are not meant for administrative or user actions. These are completely pinot internal operations.
> The client side of this call is in ServerSegmentProtocolCompletionHandler.
> 
> If you can make these changes in another PR, that will help reviewing a bit. If I miss something and it causes realtime segments not to complete segments, that will be really bad.

`ServerSegmentProtocolCompletionHandler` already has an `_authToken`. Imo the most important thing right now is getting out integration tests - batch is done, realtime is in the works.

mcvsubbu commented:
> > > > Do you need to change LLCSegmentCompletionHandlers to pay attention to the auth token?
> > > 
> > > 
> > > Probably. I'd appreciate your input on where this actually talks to controller APIs. I just noticed that this whole set of endpoints was excluded from my checklist (via swagger)
> > 
> > 
> > These endpoints should not appear in Swagger because they are used by pinot-server to commit realtime segments using the segment commit protocol. They are not meant for administrative or user actions. These are completely pinot internal operations.
> > The client side of this call is in ServerSegmentProtocolCompletionHandler.
> > If you can make these changes in another PR, that will help reviewing a bit. If I miss something and it causes realtime segments not to complete segments, that will be really bad.
> 
> `ServerSegmentProtocolCompletionHandler` already has an `_authToken`. Imo the most important thing right now is getting out integration tests - batch is done, realtime is in the works.

Working the entire realtime flow inside an integration test is definitely useful. Where I am afraid things ay break is if servers are running older version and controllers newer, or vice versa. Realtime segment upload failure will block realtime consumption. If you can ascertain that as long as auth tokens are not configured anywhere the behavior is the same that would be awesome.

apucher commented:
> Working the entire realtime flow inside an integration test is definitely useful. Where I am afraid things ay break is if servers are running older version and controllers newer, or vice versa. Realtime segment upload failure will block realtime consumption. If you can ascertain that as long as auth tokens are not configured anywhere the behavior is the same that would be awesome.

Fortunately, there aren't any logic changes here. Any auth tokens are passed via http headers and are therefore fully backwards-compatible.

The only way the segment upload can break is if controllers are configured to expect auth headers, while servers don't provide any. This would be an operations / configuration error, however. The inverse, i.e. servers providing auth headers without controllers expecting them, will work without issues (this also enables zero-downtime upgrades to authentication).

sajjad-moradi commented on pinot-controller/src/main/java/org/apache/pinot/controller/api/access/AccessControl.java:
```diff
@@ -64,4 +67,36 @@ default boolean hasAccess(String tableName, AccessType accessType, HttpHeaders h
+  }
+
+  /**
+   * Container for authentication workflow info for the Pinot UI. May be extended by implementations.
+   *
+   * Auth workflow info hold any configuration necessary to execute a UI workflow. We currently foresee supporting NONE
+   * (auth disabled), BASIC (basic auth with username and password), and OAUTH2 (token-based workflow via external
+   * issuer)
+   */
+  class AuthWorkflowInfo {
```
At LinkedIn, we're using both token-based & certificate-based authentication. We don't have a UI for the endpoints and I'm not sure what kind of information is needed to be returned to UI other than the type of authentication. What do you have in mind?

apucher replied:
afaik LinkedIn doesn't authenticate web UI calls at all (it's not yet supported by the controller UI). cert-based auth takes place in restli's tls layer, while queries are further authenticated with tokens at the request layer.

this endpoint enables us to configure different UI auth workflows depending on the scenario. at minimum we need to support NONE to maintain backwards compatibility and BASIC to support fully authenticated clusters with this PR.

sajjad-moradi replied:
If there's only NONE and BASIC, AuthWorkflowInfo only returns the type of authentication. I still don't understand what information other than auth-type is going to be returned to UI for token or certificate based authentication. Could you elaborate on that?

apucher replied:
certificate-based auth usually takes place on the lower layer (TLS/SSL).

This endpoint purely serves to configure UI auth workflows as we broaden support:
- disable auth workflows in the UI if pinot is set up without auth
- enable a user/password workflow for basic auth
- provide issuerUrl (with optional info) for oauth2 workflows
- provide secrets/redirects for Azure AD, kerberos, ... (as it comes up)

mcvsubbu replied:
@sajjad-moradi  are you ok with this resolution?

sajjad-moradi replied:
Yes sounds good.

sajjad-moradi commented on pinot-controller/src/main/java/org/apache/pinot/controller/api/access/AccessControl.java:
```diff
@@ -26,6 +26,9 @@
 @InterfaceAudience.Public
 @InterfaceStability.Stable
 public interface AccessControl {
+  String WORKFLOW_NONE = "NONE";
+  String WORKFLOW_BASIC = "BASIC";
+  String WORKFLOW_OAUTH2 = "OAUTH2";
```
I guess we should stick to the ones that are already available in the code - ie none and basic. Whenever we add new implementation, we can add them here.
Also enum is a better choice than string.

apucher replied:
Hardcoding this in an enum would eliminate any extensibility of `AuthWorkflowInfo` by third parties

sajjad-moradi commented on pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotControllerAuthResource.java:
```diff
@@ -0,0 +1,90 @@
+  }
+
+  /**
+   * Provide the auth workflow configuration for the Pinot UI to perform user authentication. Currently supports NONE
+   * (no auth), BASIC (basic auth with username and password), and OAUTH2 (token-based via external issuer)
```
AFAIK currently there's no OAuth2 support. If that's correct, please update the comment.

apucher replied:
updated

sajjad-moradi commented on pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/BasicAuthBatchIntegrationTest.java:
```diff
@@ -0,0 +1,177 @@
+    Assert.assertTrue(responseUser.get("exceptions").get(0).get("errorCode").asInt() != 0, "must return error code");
+  }
+
+//  // TODO this endpoint should be protected once UI supports auth
+//  @Test
+//  public void testControllerGetTablesNoAuth()
+//      throws Exception {
+//    System.out.println(sendGetRequest("http://localhost:18998/tables"));
+//  }
```
I don't understand the TODO here!

apucher replied:
we'll wait for the web UI to fully support auth, then protect all the remaining endpoints (including the tables endpoint referenced here)

apucher replied:
done

apucher commented:
> I think it should be "AccessType" because AccessControl.hasAccess() takes an AccessType object as its argument. Implementations of AccessControl can't change AccessType.

I agree with this from the controller side, however, `AccessType` is a controller concept only - broker implementations may specify other types of access in an extensible way. One could imagine limiting access to PII data and similar

apucher commented on pinot-common/src/main/java/org/apache/pinot/common/utils/fetcher/SegmentFetcherFactory.java:
```diff
@@ -34,29 +34,40 @@
 
   static final String SEGMENT_FETCHER_CLASS_KEY_SUFFIX = ".class";
   private static final String PROTOCOLS_KEY = "protocols";
+  private static final String AUTH_TOKEN_KEY = "auth.token";
```
done

apucher commented on pinot-controller/src/main/java/org/apache/pinot/controller/api/access/AccessControl.java:
```diff
@@ -64,4 +67,32 @@ default boolean hasAccess(String tableName, AccessType accessType, HttpHeaders h
+      return workflow;
+    }
+
+    public void setWorkflow(String workflow) {
```
It's a POJO. While Pinot isn't consistent on this, most of them seem to prefer mutable state

mcvsubbu commented on pinot-common/src/main/java/org/apache/pinot/common/utils/fetcher/SegmentFetcherFactory.java:
```diff
@@ -34,29 +34,40 @@
 public class SegmentFetcherFactory {
-  private SegmentFetcherFactory() {
-  }
+  private final static SegmentFetcherFactory instance = new SegmentFetcherFactory();
```
```suggestion
  private final static SegmentFetcherFactory INSTANCE = new SegmentFetcherFactory();
```

apucher replied:
done

mcvsubbu reviewed (commented):
Minor comment

mcvsubbu reviewed (commented):
lgtm other than a very minor comment. I have resolved all the comments of mine. Can you also work with @sajjad-moradi  to resolve his comments as well?

thanks for your patience.

apucher commented:
> lgtm other than a very minor comment. I have resolved all the comments of mine. Can you also work with @sajjad-moradi to resolve his comments as well?
> 
> thanks for your patience.

Awesome. Thanks, Subbu. Yeah, so many changes with all the glue and tooling code.

Addressed your remaining comments, and waiting for @sajjad-moradi to circle back.

sajjad-moradi commented:
> > lgtm other than a very minor comment. I have resolved all the comments of mine. Can you also work with @sajjad-moradi to resolve his comments as well?
> > thanks for your patience.
> 
> Awesome. Thanks, Subbu. Yeah, so many changes with all the glue and tooling code.
> 
> Addressed your remaining comments, and waiting for @sajjad-moradi to circle back.

LGTM.

apucher commented:
Thank you everyone for the reviews!

I've rebased the PR and I'll wait for one last CI run before merging.
